Sources/Core/GTMSessionFetcher.m

```diff
@@ -2954,22 +2955,51 @@ - (void)invokeFetchCallbacksOnCallbackQueueWithData:(nullable NSData *)data
 
 - (void)postNotificationOnMainThreadWithName:(NSString *)noteName
                                     userInfo:(nullable NSDictionary *)userInfo {
-  dispatch_block_t postBlock = ^{
-    [[NSNotificationCenter defaultCenter] postNotificationName:noteName
-                                                        object:self
-                                                      userInfo:userInfo];
-  };
+  // Historically, the notification has been posted immediately when on the main thread already, so
+  // that is continued to not break any existing code. However, if a notification gets deferred to
+  // be posted on the main thread, then even if some other code is one the main thread, it can't
+  // post it's notification before the deferred one is posted, otherwise you could get a stop and/or
+  // callback notification *before* the start notification. So if anything gets deferred to the main
+  // thread, an immediated post on the main thread also has to get delayed to maintain the order.
+  //
+  // This could get revisited to defer all notifications, but that should probably go out in a
+  // major version bump as it could be breaking to some usages who were dependent on the ordering.
 
-  if ([NSThread isMainThread]) {
-    // Post synchronously for compatibility with older code using the fetcher.
+  GTMSessionCheckNotSynchronized(self);
 
-    // Avoid calling out to other code from inside a sync block to avoid risk
-    // of a deadlock or of recursive sync.
-    GTMSessionCheckNotSynchronized(self);
+  BOOL canPostNow = NO;
+  @synchronized(self) {
+    GTMSessionMonitorSynchronized(self);
 
-    postBlock();
-  } else {
-    dispatch_async(dispatch_get_main_queue(), postBlock);
+    if (_pendingNotifications == 0 && [NSThread isMainThread]) {
+      canPostNow = YES;
+    } else {
+      // This one will get deferred, so increase the counter for deferred notifications.
+      _pendingNotifications++;
+    }
+    if (!canPostNow) {
```

Comment:
does all of this below need to be in the @synchronized block? If looks like if you took it out you could do an if (canPostNow) {...} else {...}

Reply:
You are right, it doesn't have to be.